app/models/event.rb

```diff
@@ -44,6 +52,36 @@ def formatted_scheduled_time
     scheduled_at.to_formatted_s(:long)
   end
 
+  def is_bookable_by?(applicant)
+    applicant && is_future? && !is_booked_by?(applicant) && !is_participated_by?(applicant) && !is_full?
+  end
+
+  def is_unbookable_by?(applicant)
+    is_future? && is_booked_by?(applicant)
+  end
+
+  def is_booked_by?(applicant)
+    applicant && applicants.exists?(applicant.id)
+  end
+
+  def booked_by!(applicant)
+    Booking.create(event: self, applicant: applicant) if !is_booked_by?(applicant)
+  end
+
+  # TODO: is it worth adding :counter_cache in UserEvent model?
+  def is_full?
+    capacity && capacity == participants.count
+  end
+
+  def accept_booking!(booking)
+    UserEvent.create(event: self, participant: booking.applicant)
+    booking.destroy
+  end
+
+  def is_future?
+    scheduled_at > Time.zone.now
+  end
+
```

Comment:
Excellent work here. The methods are broken down into good sizes and have very clear, understandable names. Following the logic through these methods was a delight.